core/modules/replica/SqlRequest.cc

```diff
@@ -0,0 +1,342 @@
+
+    timer().expires_from_now(boost::posix_time::milliseconds(nextTimeIvalMsec()));
+    timer().async_wait(
+        boost::bind(
+            &SqlRequest::_awaken,
+            shared_from_base<SqlRequest>(),
+            boost::asio::placeholders::error
+        )
+    );
+}
+
+
+void SqlRequest::_awaken(boost::system::error_code const& ec) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+
+    if (isAborted(ec)) return;
+
+    // IMPORTANT: the final state is required to be tested twice. The first time
+    // it's done in order to avoid deadlock on the "in-flight" callbacks reporting
+    // their completion while the request termination is in a progress. And the second
+    // test is made after acquiring the lock to recheck the state in case if it
+    // has transitioned while acquiring the lock.
+
+    if (state() == State::FINISHED) return;
+
+    util::Lock lock(_mtx, context() + __func__);
+
+    if (state() == State::FINISHED) return;
+
+    // Serialize the Status message header and the request itself into
+    // the network buffer.
+
+    buffer()->resize();
+
+    ProtocolRequestHeader hdr;
+    hdr.set_id(id());
+    hdr.set_type(ProtocolRequestHeader::REQUEST);
+    hdr.set_management_type(ProtocolManagementRequestType::REQUEST_STATUS);
+
+    buffer()->serialize(hdr);
+
+    ProtocolRequestStatus message;
+    message.set_id(id());
+    message.set_queued_type(ProtocolQueuedRequestType::SQL);
+
+    buffer()->serialize(message);
+
+    _send(lock);
+}
+
+
+void SqlRequest::_send(util::Lock const& lock) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__);
+
+    auto self = shared_from_base<SqlRequest>();
+
+    messenger()->send<ProtocolResponseSql>(
+        worker(),
+        id(),
+        buffer(),
+        [self] (string const& id,
+                bool success,
+                ProtocolResponseSql const& response) {
+
+            self->_analyze(success,
+                           response);
+        }
+    );
+}
+
+
+void SqlRequest::_analyze(bool success,
+                          ProtocolResponseSql const& message) {
+
+    LOGS(_log, LOG_LVL_DEBUG, context() << __func__ << "  success=" << (success ? "true" : "false"));
+
+    // This method is called on behalf of an asynchronous callback fired
+    // upon a completion of the request within method _send() - the only
+    // client of _analyze(). So, we should take care of proper locking and watch
+    // for possible state transition which might occur while the async I/O was
+    // still in a progress.
+
+    // IMPORTANT: the final state is required to be tested twice. The first time
+    // it's done in order to avoid deadlock on the "in-flight" callbacks reporting
+    // their completion while the request termination is in a progress. And the second
+    // test is made after acquiring the lock to recheck the state in case if it
+    // has transitioned while acquiring the lock.
```

Comment:
State transitions in the Controller-side Request classes may occur in the following cases:
# a completion of a response from a worker
# an expiration of a request (there is an asynchronous timer started when the request starts)
# e explicit cancellation of a request by a client who might get lost his/her interest in the request
A problem is that these events may happen in any thread. Each even will progress the state of a request in one direction - towards its state `Request::FINISHED`. And, what is really important here, is that in either of those cases a notification callback would be initiated (indirectly) by calling method `Request::finish`. It's is a *REQUIREMENT* that the callback was called once and only once, regardless of a completion status of the request. What you've seen provides a guarantee for that. The first check prevents deadlocks in case if more then one completion condition (among those above mentioned) is being executed by different threads.